hyperon_das_atomdb/adapters/ram_only.py

```diff
@@ -71,53 +84,85 @@ def _add_atom_type(self, _name: str, _type: Optional[str] = 'Type'):
-        named_type_hash: str,
-        key: str,
-        targets_hash: List[str],
+        self, composite_type_hash: str, named_type_hash: str, key: str, targets_hash: List[str]
     ) -> None:
         template_composite_type_hash = self.db.templates.get(composite_type_hash)
         template_named_type_hash = self.db.templates.get(named_type_hash)
 
         if template_composite_type_hash is not None:
-            # template_composite_type_hash.append([key, targets_hash])
             template_composite_type_hash.append((key, tuple(targets_hash)))
         else:
-            # self.db.templates[composite_type_hash] = [[key, targets_hash]]
             self.db.templates[composite_type_hash] = [(key, tuple(targets_hash))]
 
         if template_named_type_hash is not None:
-            # template_named_type_hash.append([key, targets_hash])
             template_named_type_hash.append((key, tuple(targets_hash)))
         else:
-            # self.db.templates[named_type_hash] = [[key, targets_hash]]
             self.db.templates[named_type_hash] = [(key, tuple(targets_hash))]
 
+    def _delete_templates(self, link_document: dict, targets_hash: List[str]) -> None:
+        template_composite_type = self.db.templates.get(link_document['composite_type_hash'], [])
+        if len(template_composite_type) > 0:
+            template_composite_type.remove((link_document['_id'], tuple(targets_hash)))
+        if len(template_composite_type) == 0:
+            self.db.templates.pop(link_document['composite_type_hash'], None)
+
+        template_named_type = self.db.templates.get(link_document['named_type_hash'], [])
+        if len(template_named_type) > 0:
+            template_named_type.remove((link_document['_id'], tuple(targets_hash)))
+        if len(template_named_type) == 0:
+            self.db.templates.pop(link_document['named_type_hash'], None)
+
     def _add_patterns(self, named_type_hash: str, key: str, targets_hash: List[str]):
         pattern_keys = build_patern_keys([named_type_hash, *targets_hash])
 
         for pattern_key in pattern_keys:
             pattern_key_hash = self.db.patterns.get(pattern_key)
             if pattern_key_hash is not None:
-                # pattern_key_hash.append([key, targets_hash])
                 pattern_key_hash.append((key, tuple(targets_hash)))
             else:
-                # self.db.patterns[pattern_key] = [[key, targets_hash]]
                 self.db.patterns[pattern_key] = [(key, tuple(targets_hash))]
 
+    def _delete_patterns(self, link_document: dict, targets_hash: List[str]) -> None:
+        pattern_keys = build_patern_keys([link_document['named_type_hash'], *targets_hash])
+        for pattern_key in pattern_keys:
+            pattern = self.db.patterns.get(pattern_key, [])
+            if len(pattern) > 0:
+                pattern.remove((link_document['_id'], tuple(targets_hash)))
+            if len(pattern) == 0:
+                self.db.patterns.pop(pattern_key, None)
```

Comment:
This is wrong. The pattern key is still searchable even if there's no atom that matches it. We should leave the empty list.

Reply:
Sorry, I couldn't understand your point. After removing an item from a `pattern_key` I do another check and if this list is empty I completely remove `pattern_key`

Reply:
`pattern` holds a list of (handle, targets) that matched the pattern identified as `pattern_key`. In line 159 you check if this list became empty after removing the entry related to the link you are deleting (which you did in line 158). If this list is empty, you remove `pattern_key` out of `self.db.patterns` in line 160. This is wrong, you should keep the empty list there.

Reply:
Why do we need to keep this `pattern` empty? We talked about this case a few days ago and agreed that if the list is empty the patterns should be deleted

Reply:
I'll take back my comment regarding the correctness of the code. It's correct, indeed, as any search for non-existent keys (either here in ramOnly and in redis-mongo) will return an empty list anyway. However it's still unnecessary. Perhaps I misunderstood the context when we discussed this before. Anyway I see no point in worrying about removing empty lists (in this and in all the other similar methods).

If you think we need to remove them, please provide details on why you think we should do it. If it's just to keep the tables clean of empty values (which would return [] if queried from non-existent keys anyway) I vote for just adding another "cleanup" task in the card https://github.com/singnet/das-atom-db/issues/90

Reply:
done!